src/lss.jl

```diff
@@ -427,3 +427,158 @@ function __solve(prob::ForwardLSSProblem,sensealg::ForwardLSS,alpha::Cos2Windowi
+  isinplace = DiffEqBase.isinplace(f)
+
+  p == nothing && error("You must have parameters to use parameter sensitivity calculations!")
+  !(sol.u isa AbstractVector) && error("`u` has to be an AbstractVector.")
+
+
+  sense = LSSSensitivityFunction(sensealg,f,f.analytic,f.jac,
+                                     f.jac_prototype,f.sparsity,f.paramjac,
+                                     u0,sensealg,
+                                     p,similar(u0),f.mass_matrix,
+                                     f.colorvec,
+                                     tspan,g,dg)
+
+  @unpack numparams, numindvar = sense
+  Nt = length(sol.t)
+  Ndt = Nt-one(Nt)
+
+  # pre-allocate variables
+  dt = similar(sol.t, Ndt)
+  umid = Matrix{eltype(u0)}(undef,numindvar,Ndt)
+  dudt = Matrix{eltype(u0)}(undef,numindvar,Ndt)
+  # compute their values
+  discretize_ref_trajectory!(dt, umid, dudt, sol, Ndt)
+
+  S = LSSSchur(dt,u0,numindvar,Nt,Ndt,sensealg.alpha)
+
+  if sensealg.alpha isa Number
+    g0 = g(u0,p,tspan[1])
+  else
+    g0 = nothing
+  end
+
+  b = Vector{eltype(u0)}(undef,numindvar*Ndt)
+  h = Vector{eltype(u0)}(undef,Ndt)
+  wa = similar(dt,numindvar*Ndt)
+
+  Δt = tspan[2] - tspan[1]
+  wB!(S,Δt,Nt,numindvar,dt)
+  wE!(S,Δt,dt,sensealg.alpha)
+
+  B!(S,dt,umid,sense,sensealg)
+  E!(S,dudt,sensealg.alpha)
+  S!(S)
+  wBcorrect!(S,sol,g,Nt,sense,sensealg)
+
+  h!(h,g0,g,umid,p,S.wEinv)
+
+  res = similar(u0, numparams)
+
+  AdjointLSSProblem{typeof(sensealg),typeof(sense),typeof(sol),typeof(dt),
+    typeof(umid),typeof(dudt),
+    typeof(S),typeof(h),typeof(b),typeof(wa),typeof(Δt),
+    typeof(g0),typeof(g),typeof(res)}(sensealg,sense,sol,dt,umid,dudt,S,h,b,wa,Δt,Nt,g0,g,
+    res)
+end
+
+function h!(h,g0,g,u,p,wEinv)
+
+  for (j,uj) in enumerate(eachcol(u))
+    # compute objective
+    h[j] = g(uj,p,nothing)
+  end
+  h .= -(h .- mean(h)) / (size(u)[2])
+
+  @. h = wEinv*h
+
+  return nothing
+end
+
+function wBcorrect!(S,sol,g,Nt,sense,sensealg)
+  @unpack dg_val, pgpu, pgpu_config, numparams, numindvar, uf = sense
+  @unpack wBinv = S
+
+  for (i,u) in enumerate(sol.u)
+    _wBinv = @view wBinv[(i-1)*numindvar+1:i*numindvar]
+    if dg_val isa Tuple
+      DiffEqSensitivity.gradient!(dg_val[1], pgpu, u, sensealg, pgpu_config)
+      @. _wBinv = _wBinv*dg_val[1]/Nt
+    else
+      DiffEqSensitivity.gradient!(dg_val, pgpu, u, sensealg, pgpu_config)
+      @. _wBinv = _wBinv*dg_val/Nt
+    end
+  end
+  return nothing
+end
+
+function __solve(prob::AdjointLSSProblem; t0skip=zero(prob.Δt), t1skip=zero(prob.Δt))
+  __solve(prob,prob.sensealg,prob.sensealg.alpha,t0skip,t1skip)
+end
+
+function __solve(prob::AdjointLSSProblem,sensealg::AdjointLSS,alpha::Number,t0skip,t1skip)
+  @unpack sol, S, Δt, diffcache, h, b, wa, res, g, g0, umid = prob
+  @unpack wBinv, B, E, Smat = S
+  @unpack dg_val, pgpp, pgpp_config, numparams, numindvar, uf, f, f_cache, pJ, pf, paramjac_config = diffcache
+
+  b .= E*h + B*wBinv
+  wa .= Smat\b
```

Comment:
pre-caching the Smat factorization here?

Reply:
Should this go to `AdjointLSSProblem` then? In this case `Smat\b` is only called once.

Reply:
yeah, just `lu!(Smat)` before saving it.

Reply:
is this the same as forward LSS? If it's not, we should probably get that updated. I forgot about `lu!` for that 😅

Reply:
yup I updated it for both. `lu!(Smat)` just returned the `LU` decomposition though (i.e., `Smat` stayed the same), so I went for removing `Smat` from the `LSSSchur` struct and instead added a field storing the LU decomposition to the respective LSSProblem.